File: backend/app/graphql/mutations/sign_out.rb
module Mutations
  class SignOut < Mutations::BaseMutation
    graphql_name "SignOut"

    argument :token, String, required: true

    field :user, Types::UserType, null: false

    def resolve(args)
      user = User.find_by(authentication_token: args[:token])
      success = user.present? && user.reset_authentication_token!

      MutationResult.call(
        success: success,
        errors: [],
      )
    rescue
      GraphQL::ExecutionError.new("Could not sign out user")
    end
  end
end

File: frontend/src/mutations/signOut.js
import gql from 'graphql-tag';

const mutation = gql`
  mutation signOut($token: String!) {
    signOut(input: { token: $token }) {
      success
      errors
    }
  }
`;

export default function signOut({
  apollo,
  token,
}) {
  return apollo.mutate({
    mutation,
    variables: {
      token,
    },
  });
}
